ddl/ddl_api.go

```diff
@@ -4305,57 +4306,124 @@ func (d *ddl) TruncateTable(ctx sessionctx.Context, ti ast.Ident) error {
+	err = d.callHookOnChanged(err)
+	return errors.Trace(err)
+}
+
+func extractTblInfos(is infoschema.InfoSchema, oldIdent, newIdent ast.Ident, isAlterTable bool, tables map[string]int64) ([]*model.DBInfo, int64, error) {
 	oldSchema, ok := is.SchemaByName(oldIdent.Schema)
+	newIdentKey := getIdentKey(newIdent)
 	if !ok {
 		if isAlterTable {
-			return infoschema.ErrTableNotExists.GenWithStackByArgs(oldIdent.Schema, oldIdent.Name)
+			return nil, 0, infoschema.ErrTableNotExists.GenWithStackByArgs(oldIdent.Schema, oldIdent.Name)
 		}
-		if is.TableExists(newIdent.Schema, newIdent.Name) {
-			return infoschema.ErrTableExists.GenWithStackByArgs(newIdent)
+		tableID, ok := tables[newIdentKey]
+		if (ok && tableID != tableNotExist) || is.TableExists(newIdent.Schema, newIdent.Name) {
+			return nil, 0, infoschema.ErrTableExists.GenWithStackByArgs(newIdent)
 		}
-		return errFileNotFound.GenWithStackByArgs(oldIdent.Schema, oldIdent.Name)
+		return nil, 0, errFileNotFound.GenWithStackByArgs(oldIdent.Schema, oldIdent.Name)
 	}
-	oldTbl, err := is.TableByName(oldIdent.Schema, oldIdent.Name)
-	if err != nil {
-		if isAlterTable {
-			return infoschema.ErrTableNotExists.GenWithStackByArgs(oldIdent.Schema, oldIdent.Name)
-		}
-		if is.TableExists(newIdent.Schema, newIdent.Name) {
-			return infoschema.ErrTableExists.GenWithStackByArgs(newIdent)
+	oldIdentKey := getIdentKey(oldIdent)
+	oldTableID, ok := tables[oldIdentKey]
+	var oldTbl table.Table
+	var err error
+	if (ok && oldTableID == tableNotExist) || !ok {
+		oldTbl, err = is.TableByName(oldIdent.Schema, oldIdent.Name)
+		if (ok && oldTableID == tableNotExist) || err != nil {
+			if isAlterTable {
+				return nil, 0, infoschema.ErrTableNotExists.GenWithStackByArgs(oldIdent.Schema, oldIdent.Name)
+			}
+			if is.TableExists(newIdent.Schema, newIdent.Name) {
+				return nil, 0, infoschema.ErrTableExists.GenWithStackByArgs(newIdent)
+			}
+			return nil, 0, errFileNotFound.GenWithStackByArgs(oldIdent.Schema, oldIdent.Name)
 		}
-		return errFileNotFound.GenWithStackByArgs(oldIdent.Schema, oldIdent.Name)
+		oldTableID = oldTbl.Meta().ID
 	}
 	if isAlterTable && newIdent.Schema.L == oldIdent.Schema.L && newIdent.Name.L == oldIdent.Name.L {
-		// oldIdent is equal to newIdent, do nothing
-		return nil
+		//oldIdent is equal to newIdent, do nothing
+		return nil, 0, nil
 	}
 	newSchema, ok := is.SchemaByName(newIdent.Schema)
 	if !ok {
-		return ErrErrorOnRename.GenWithStackByArgs(
+		return nil, 0, ErrErrorOnRename.GenWithStackByArgs(
 			fmt.Sprintf("%s.%s", oldIdent.Schema, oldIdent.Name),
 			fmt.Sprintf("%s.%s", newIdent.Schema, newIdent.Name),
 			168,
 			fmt.Sprintf("Database `%s` doesn't exist", newIdent.Schema))
 	}
-	if is.TableExists(newIdent.Schema, newIdent.Name) {
-		return infoschema.ErrTableExists.GenWithStackByArgs(newIdent)
+
+	newTableID, ok := tables[newIdentKey]
+	if (ok && newTableID != tableNotExist) || (!ok && is.TableExists(newIdent.Schema, newIdent.Name)) {
```

Comment:
The code is redundant. You can extract the judgement into 2 private functions: tableExists and tableNotExists.